gustaf/io/meshio.py

```diff
@@ -18,58 +19,96 @@
-    Currently cannot process boundary.
 
-    Note
-    -----
-    This is more or less a direct copy from the original gustav implementation.
-    A lot of the meshio information are lost. When boundaries and multi-patch
-    definitions are added this needs to be revisited and extended.
+    The function reads all gustaf-processable element types from meshio, which
+    are sorted by the dimensionality of their element types. Different
+    elements can occur at the same dimensionality, e.g., in mixed meshes,
+    or as sub-topologies, e.g., as boundaries or interfaces.
+
+    If set_boundary is set true, all (d-1)-dimensional submesh node identifiers
+    are assigned to MESH_TYPES.BC without check.
+
+    Return_only_one_mesh ensures compatibility with previous gustaf versions.
+
+    Loading physical groups is not supported yet.
 
     Parameters
     ------------
     fname: str | pathlib.Path
+       Input filename
+    set_boundary : bool, optional, default is False
+       Assign nodes of lower-dimensional meshes mesh.BC
+    return_only_one_mesh : bool, optional, default is True
+        Return only the highest-dimensional mesh, ensures compartibility.
 
     Returns
     --------
-    MESH_TYPES
+    MESH_TYPES or list(MESH_TYPES)
     """
-    mesh_type = Faces
-
     fname = pathlib.Path(fname)
     if not (fname.exists() and fname.is_file()):
         raise ValueError(
             "The given file does not point to file. The given path is: "
             f"{fname.resolve()}"
         )
 
+    # Read mesh
     meshio_mesh: meshio.Mesh = meshio.read(fname)
     vertices = meshio_mesh.points
 
-    # check for 2D mesh
-    # Try for triangle grid
-    cells = meshio_mesh.get_cells_type("triangle")
-    # If  no triangle elements, try for square
-    if len(cells) == 0:
-        cells = meshio_mesh.get_cells_type("quad")
-    if not len(cells) > 0:
-        # 3D mesh
-        mesh_type = Volumes
-        cells = meshio_mesh.get_cells_type("tetra")
-        if len(cells) == 0:
-            cells = meshio_mesh.get_cells_type("hexahedron")
-
-    for i, cell in enumerate(cells):
-        if i == 0:
-            elements = cell.data
-        else:
-            elements = np.vstack((elements, cell.data))
-
-    mesh = mesh_type(vertices=vertices, elements=elements)
+    # Maps meshio element types to gustaf types and dimensionality
+    # Meshio vertex elements are ommited, as they do not follow gustaf logics
+    meshio_mapping = {
+        "hexahedron": [Volumes, 3],
+        "tetra": [Volumes, 3],
+        "quad": [Faces, 2],
+        "triangle": [Faces, 2],
+        "line": [Edges, 1],
+    }
+    meshes = [
+        [
+            meshio_mapping[key][0](
+                vertices=vertices, elements=meshio_mesh.cells_dict[key]
+            ),
+            meshio_mapping[key][1],
+        ]
+        for key in meshio_mapping.keys()
+        if key in meshio_mesh.cells_dict.keys()
+    ]
```

Comment:
```suggestion
    meshes = [
        [
            meshio_mapping[key][0](
                vertices=vertices, elements=meshio_mesh.cells_dict[key]
            ),
            meshio_mapping[key][1],
        ]
        for key in meshio_mapping.keys()
        if key in meshio_mesh.cells_dict.keys()
    ]
```
can you please just write a normal for loop here?

Reply:
this is not resolved